# Albums and Songs Lab

### Introduction

In this lesson, we'll use the skills we have learned over the past several lessons to answer questions about the top songs, artists and albums over the past fifty years.

### Working with Songs

Let's start by working with data regarding top 500 albums according to the Rolling Stone Magazine.

In [ ]:
import pandas as pd
url = "https://raw.githubusercontent.com/eng-6-22/mod-1-a-data-structures/master/6-top-songs/data.csv"
df = pd.read_csv(url)
albums = df.to_dict('records')

In [ ]:
# albums[:2]

In [ ]:
len(albums)

478

> Well, 478.

Let's write some functions to help us better explore the data.

* `all_albums` - Takes an argument of albums and returns the list of album names.

* `all_artists` - Takes argument of list of albums and returns a list of all artists (where each element is a string), and no artist is repeated. 

* `find_by_name` - Has one argument of `album_name`. Returns a dictionary of the correct album, or `None` if no album is found.

* `find_by_ranks` - Takes `begin_rank` and `end_rank` as arguments.  Also possible to execute the function by just providing the `begin_rank` or `end_rank` (and not both).  If no arguments are provided the entire list of albums are returned.

* `find_by_years` - Takes `begin_year` and `end_year` as arguments, and returns a list of dictionaries for albums between those years.  Also possible to execute the function by just providing the `begin_year` or `end_year` (and not both).

### Working with Songs

Next, let's load up data related to songs, and data that connects albums and songs.

In [2]:
import pandas as pd
songs_url = "https://raw.githubusercontent.com/eng-6-22/mod-1-a-data-structures/master/6-top-songs/top-500-songs.txt"
songs_df = pd.read_csv(songs_url, sep='\t', header = None, names = ['rank', 'song', 'artist', 'year'])
songs = songs_df.to_dict('records')

track_url = "https://raw.githubusercontent.com/eng-6-22/mod-1-a-data-structures/master/6-top-songs/track_data.json"
albums_and_tracks = pd.read_json(track_url)
albums_tracks = albums_and_tracks.to_dict('records')

In [3]:
songs[:5]

[{'artist': 'Bob Dylan',
  'rank': 1,
  'song': 'Like a Rolling Stone',
  'year': 1965},
 {'artist': 'The Rolling Stones',
  'rank': 2,
  'song': 'Satisfaction',
  'year': 1965},
 {'artist': 'John Lennon', 'rank': 3, 'song': 'Imagine', 'year': 1971},
 {'artist': 'Marvin Gaye', 'rank': 4, 'song': "What's Going On", 'year': 1971},
 {'artist': 'Aretha Franklin', 'rank': 5, 'song': 'Respect', 'year': 1967}]

In [4]:
albums_tracks[0]

{'album': "Sgt. Pepper's Lonely Hearts Club Band",
 'artist': 'The Beatles',
 'tracks': ["Sgt. Pepper's Lonely Hearts Club Band - Remix",
  'With A Little Help From My Friends - Remix',
  'Lucy In The Sky With Diamonds - Remix',
  'Getting Better - Remix',
  'Fixing A Hole - Remix',
  "She's Leaving Home - Remix",
  'Being For The Benefit Of Mr. Kite! - Remix',
  'Within You Without You - Remix',
  "When I'm Sixty-Four - Remix",
  'Lovely Rita - Remix',
  'Good Morning Good Morning - Remix',
  "Sgt. Pepper's Lonely Hearts Club Band (Reprise) - Remix",
  'A Day In The Life - Remix',
  "Sgt. Pepper's Lonely Hearts Club Band - Take 9 And Speech",
  'With A Little Help From My Friends - Take 1 / False Start And Take 2 / Instrumental',
  'Lucy In The Sky With Diamonds - Take 1',
  'Getting Better - Take 1 / Instrumental And Speech At The End',
  'Fixing A Hole - Speech And Take 3',
  "She's Leaving Home - Take 1 / Instrumental",
  'Being For The Benefit Of Mr. Kite! - Take 4',
  'Within You

* Write functions that perform the following: 

`album_most_top_songs` - 
    * Returns the name of the artist and album that has that most songs featured on the top 500 songs list

`top_ten_albums_by_songs` - returns a dictionary with the 10 albums that have the most songs that appear in the top songs list. The album names should be the keys and the corresponding values should be the number of songs that appear on the top 500 list.

In [74]:
tracks_uniq = []
for album_track in albums_tracks:
  for track_name in album_track['tracks']:
    cleaned_name = track_name.split(' - ')[0]
    track = {'track': cleaned_name, 'album': album_track['album'], 'artist': album_track['artist']}
    tracks_uniq.append(track)


tracks_uniq[:3]

[{'album': "Sgt. Pepper's Lonely Hearts Club Band",
  'artist': 'The Beatles',
  'track': "Sgt. Pepper's Lonely Hearts Club Band"},
 {'album': "Sgt. Pepper's Lonely Hearts Club Band",
  'artist': 'The Beatles',
  'track': 'With A Little Help From My Friends'},
 {'album': "Sgt. Pepper's Lonely Hearts Club Band",
  'artist': 'The Beatles',
  'track': 'Lucy In The Sky With Diamonds'}]

In [84]:
tracks = []
for album_track in albums_tracks:
  for track_name in album_track['tracks']:
    cleaned_name = track_name.split(' - ')[0]
    track = {'track': cleaned_name, 'album': album_track['album'], 'artist': album_track['artist']}
    if track not in tracks:
      tracks.append(track)

tracks[:3]

[{'album': "Sgt. Pepper's Lonely Hearts Club Band",
  'artist': 'The Beatles',
  'track': "Sgt. Pepper's Lonely Hearts Club Band"},
 {'album': "Sgt. Pepper's Lonely Hearts Club Band",
  'artist': 'The Beatles',
  'track': 'With A Little Help From My Friends'},
 {'album': "Sgt. Pepper's Lonely Hearts Club Band",
  'artist': 'The Beatles',
  'track': 'Lucy In The Sky With Diamonds'}]

In [77]:
def rank_song(track):
  for song in songs:
    if song['song'] == track:
      return int(song['rank'])
  return 501

rank_song('Satisfaction')

2

In [78]:
tracks_with_rank = []
for track in tracks:
  track_with_rank = {'track': track['track'], 'album': track['album'], 'artist': track['artist'], 'rank': rank_song(track['track'])}
  tracks_with_rank.append(track_with_rank)      

tracks_with_rank[-1]    

{'album': 'Aquemini', 'artist': 'OutKast', 'rank': 501, 'track': 'Chonkyfire'}

In [87]:
top_songs = sorted(tracks_with_rank, key=lambda x: x['rank'])
top_songs[-1]


{'album': 'Aquemini', 'artist': 'OutKast', 'rank': 501, 'track': 'Chonkyfire'}

In [ ]:
top_artist_album = {}
for song in top_songs:
  if song['rank'] <= 500:
    top_artist_album[(song['album'],song )]